jbmc/src/java_bytecode/remove_instanceof.cpp

```diff
@@ -94,30 +131,38 @@ bool remove_instanceoft::lower_instanceof(
     "instanceof second operand should be a type");
   const typet &target_type=target_arg.type();
 
-  // Find all types we know about that satisfy the given requirement:
   INVARIANT(
     target_type.id() == ID_struct_tag,
     "instanceof second operand should have a simple type");
-  const irep_idt &target_name =
-    to_struct_tag_type(target_type).get_identifier();
-  std::vector<irep_idt> children=
-    class_hierarchy.get_children_trans(target_name);
-  children.push_back(target_name);
-  // Sort alphabetically to make order of generated disjuncts
-  // independent of class loading order
-  std::sort(
-    children.begin(), children.end(), [](const irep_idt &a, const irep_idt &b) {
-      return a.compare(b) < 0;
-    });
 
-  auto jlo = to_struct_tag_type(java_lang_object_type().subtype());
-  exprt object_clsid = get_class_identifier_field(check_ptr, jlo, ns);
+  // If checking for an array type, the target type must have the same
+  // dimension:
+  const auto underlying_type_and_dimension =
+    java_reference_array_dimension_and_element_type(
+      to_struct_tag_type(target_type));
 
-  // Replace the instanceof operation with (check_ptr != null &&
-  //  (check_ptr->@class_identifier == "A" ||
-  //   check_ptr->@class_identifier == "B" || ...
+  const irep_idt &underlying_name =
+    to_struct_tag_type(underlying_type_and_dimension.first).get_identifier();
 
-  // By operating directly on a pointer rather than using a temporary
+  auto jlo = to_struct_tag_type(java_lang_object_type().subtype());
+  bool underlying_type_is_object = underlying_type_and_dimension.first == jlo;
+  bool target_type_is_array = underlying_type_and_dimension.second > 0;
```

Comment:
:no_entry_sign: Perhaps re-reviewing too soon - but this variable should definitely be called `target_type_is_reference_array` given the dimension is zero for primitive arrays